pkg/service/async_command_handler.go

```diff
@@ -0,0 +1,129 @@
+	return ac.serviceMessage.Headers()
+}
+
+// ParsedCommand returns the command parsed in the specified type
+func (ac *AsyncCommandMessage) ParsedCommand() message.AsyncCommand {
+	return ac.parsedCommand
+}
+
+// NotifyFailed will notify a failure in the handling of the control message
+// If err == nil, this acts the same as NotifySuccess
+func (ac *AsyncCommandMessage) NotifyFailed(err error) {
+	if err == nil {
+		ac.NotifySuccess()
+		return
+	}
+
+	commandResult := message.AsyncCommandResult{
+		CommandId: ac.parsedCommand.SerializedId(),
+		Error:     err.Error(),
+	}
+
+	if err := ac.service.SendAndWaitForAck(ac.resultOpCode, commandResult); err != nil {
+		logging.FromContext(ac.ctx).Errorw("Error while propagating the async command result", zap.Any("asyncCommandResult", commandResult), zap.Error(err))
+	}
+}
+
+// NotifySuccess will notify the success of the handling of the control message
+func (ac *AsyncCommandMessage) NotifySuccess() {
+	commandResult := message.AsyncCommandResult{
+		CommandId: ac.parsedCommand.SerializedId(),
+	}
+
+	if err := ac.service.SendAndWaitForAck(ac.resultOpCode, commandResult); err != nil {
+		logging.FromContext(ac.ctx).Errorw("Error while propagating the async command result", zap.Any("asyncCommandResult", commandResult), zap.Error(err))
+	}
+}
+
+type asyncCommandHandler struct {
+	svc          control.Service
+	resultOpCode control.OpCode
+
+	asyncCommandType reflect.Type
+
+	handler func(context.Context, AsyncCommandMessage)
+}
+
+// NewAsyncCommandHandler returns a control.MessageHandler that wraps the provided handler, but passing an AsyncCommandMessage.
+// This handler automatically parses the message payload using the provided type, and using the AsyncCommandMessage it notifies the result of the command to the sender.
+func NewAsyncCommandHandler(svc control.Service, asyncCommandType message.AsyncCommand, resultOpCode control.OpCode, handler func(context.Context, AsyncCommandMessage)) control.MessageHandler {
+	return &asyncCommandHandler{
+		svc:              svc,
+		resultOpCode:     resultOpCode,
+		asyncCommandType: reflect.Indirect(reflect.ValueOf(asyncCommandType)).Type(),
+		handler:          handler,
+	}
+}
+
+func (f *asyncCommandHandler) HandleServiceMessage(ctx context.Context, msg control.ServiceMessage) {
+	// This is a pointer to asyncCommandType
+	commandValue := reflect.New(f.asyncCommandType).Interface()
+
+	// This assertion is always true b/c asyncCommandType is per contract a message.AsyncCommand,
+	// which implements encoding.BinaryUnmarshaler
+	err := commandValue.(encoding.BinaryUnmarshaler).UnmarshalBinary(msg.Payload())
+	if err != nil {
+		logging.FromContext(ctx).Warnw("Error while parsing the async command", zap.Error(err))
+		msg.AckWithError(fmt.Errorf("error while parsing the async command: %w", err))
+		return
+	}
+
+	asyncCommandMsg := AsyncCommandMessage{
+		serviceMessage: msg,
+		parsedCommand:  commandValue.(message.AsyncCommand),
```

Comment:
why `parsedCommand` ?  isn't that just the command (or command code), of the message?

Reply:
nope, renamed the vars to make it clear